unittests/test_pipeline.py

```diff
@@ -1249,6 +1249,134 @@ def extract_perf(patt, tag):
     assert 'v3' in log_output
 
 
+@pytest.fixture
+def perftest(testsys_system, perf_file, sanity_file):
+    class MyTest(rfm.RunOnlyRegressionTest):
+        sourcesdir = None
+
+        @sanity_function
+        def dummy_sanity(self):
+            return sn.assert_found(r'success', sanity_file)
+
+        @performance_function('unit')
+        def value1(self):
+            pass
+
+        @performance_function('unit')
+        def value2(self):
+            pass
+
+        @performance_function('unit', perf_key='value3')
+        def value_3(self):
+            pass
+
+    yield MyTest()
+
+
+def test_validate_default_perf_variables(perftest):
+    assert len(perftest.perf_variables) == 3
+    assert 'value1' in perftest.perf_variables
+    assert 'value2' in perftest.perf_variables
+    assert 'value3' in perftest.perf_variables
+
+
+def test_perf_vars_without_reference(perftest, sanity_file,
+                                     perf_file, dummy_gpu_exec_ctx):
+    logfile = 'perf.log'
+
+    @sn.deferrable
+    def extract_perf(patt, tag):
+        val = sn.evaluate(
+            sn.extractsingle(patt, perf_file, tag, float)
+        )
+        with open(logfile, 'a') as fp:
+            fp.write(f'{tag}={val}')
+
+        return val
+
+    sanity_file.write_text('result = success\n')
+    perf_file.write_text('perf1 = 1.0\n'
+                         'perf3 = 3.3\n')
+    perftest.perf_variables = {
+        'value1': sn.make_performance_function(
+            extract_perf(r'perf1 = (?P<v1>\S+)', 'v1'), 'unit'
+        ),
+        'value3': sn.make_performance_function(
+            extract_perf, 'unit', r'perf3 = (?P<v3>\S+)', 'v3'
+        )
+    }
+    _run_sanity(perftest, *dummy_gpu_exec_ctx)
+
+    logfile = os.path.join(perftest.stagedir, logfile)
+    with open(logfile) as fp:
+        log_output = fp.read()
+
+    assert 'v1' in log_output
+    assert 'v3' in log_output
+
+
+def test_perf_vars_with_reference(perftest, sanity_file,
+                                  perf_file, dummy_gpu_exec_ctx):
+    logfile = 'perf.log'
+
+    @sn.deferrable
+    def extract_perf(patt, tag):
+        val = sn.evaluate(
+            sn.extractsingle(patt, perf_file, tag, float)
+        )
+        with open(logfile, 'a') as fp:
+            fp.write(f'{tag}={val}')
+
+        return val
+
+    sanity_file.write_text('result = success\n')
+    perf_file.write_text('perf1 = 1.0\n')
+
+    # Make the unit in the reference different from the performance function
+    perftest.reference = {
+        '*': {
+            'value1': (0, None, None, 'unit_')
```

Comment:
I think that you could assert that the warning about the unit mismatch in `log_output`.

Reply:
You're right. You can't! I thought we were using the logger 😅